### ISSUE TITLE
Tame hyper-polymorphic types

### ISSUE THREAD

**Issue by joseandro** [MinaProtocol/mina#12064]
Lots of types `t` are declared as some form of records whose field types are parameters of said type `t`. This allows genericity, which is good, but also too much of it since possible type constraints over field types cannot be enforced, AFAIK, in OCaml's type system.

Another drawback is clear when there are lots of type parameters (e.g., more than 3): the declaration becomes simply very hard to understand.

One way out of it is to keep the generic type within the implementation but force the clients of the API to instances (e.g., through functors) of this generic type.

But maybe this type is actually not generic at all and there is only one instance of it. In that case, I propose to remove the genericity, at least in the API (`mli`).

In summary, this would mean, when possible:

- **one instance:** keep the generic type but only exposes the monomorphic instance
- **2+ instances:** keep the generic type and expose a way to instanciate it for
clients of the module.

This should work even better in conjunction with #12062
